### PR TITLE
Implement DistribtuedReadingService

### PR DESCRIPTION
**Pull request by ejguan**
Add DistributedReadingService
- Single process
- Share shuffle seeds across distributed process
- Automatically distributed sharding

Add tests for both `DataLoader2` and `DataLoader`.
- Spawn processes
- Elastic training

### CODE REVIEW

File: torchdata/dataloader2/reading_service.py [hunk 1]
```diff
@@ -235,3 +240,84 @@ def finalize(self) -> None:
...
+        return datapipe
+
+    def initialize_iteration(self) -> None:
+        seed = self._share_seed()
+        self._seed_generator.manual_seed(seed)
+        assert self._datapipe is not None
+        self._datapipe = torch.utils.data.graph_settings.apply_shuffle_seed(
+            self._datapipe,
+            self._seed_generator,
+        )
```
Comment (ejguan) [inline torchdata/dataloader2/reading_service.py:277]: This should be changed after `initialize_iteration` is modified.

**Comment by VitalyFedyunin** [comment]
Just to be clear, does the first version assumes 1 process per rank?

File: torchdata/dataloader2/reading_service.py (hunk 1 above)
Comment (NivekT) [reply torchdata/dataloader2/reading_service.py:277]: I may be out of the loop but what is the planned change here?

**Review by NivekT — COMMENTED** [review]
Question about the distributed design in general:
1. Are we expecting users to always use something like `torch.multiprocessing.spawn` to start distributed training? And that will properly start/clean up all the processes?
2. To what extent is the optimization from #555 compatible with this? Maybe it can work for every node?

**Comment by ejguan** [comment]
> Just to be clear, does the first version assumes 1 process per rank?

Yeah. I think our next step is to support mixed reading service (distributed reading service + multiprocessing reading service)

**Comment by ejguan** [comment]
> 1. Are we expecting users to always use something like `torch.multiprocessing.spawn` to start distributed training? And that will properly start/clean up all the processes?

Nope. I can add a different test for elastic training.

File: torchdata/dataloader2/reading_service.py (hunk 1 above)
Comment (ejguan) [reply torchdata/dataloader2/reading_service.py:277]: `initialize_iteration` should accept the RNG from `DataLoader2` to set randomness over the pipeline within RS.

File: test/test_distributed.py [hunk 2]
```diff
@@ -15,6 +15,7 @@
 import torch.multiprocessing as mp
 from torch.testing._internal.common_utils import instantiate_parametrized_tests, parametrize
 
+from torchdata.dataloader2 import DataLoader2, DistributedReadingService
 from torchdata.datapipes.iter import IterableWrapper
 from torchdata.datapipes.iter.util.prefetch import PrefetchTimeoutError
```
[the remark this answers is not in the document]
Comment (ejguan) [reply test/test_distributed.py:21]: Oh, yes. Thanks for catching it. Will added it in this PR!

**Comment by ejguan** [comment]
This PR should be ready to review. cc: @NivekT @VitalyFedyunin 
In terms of multiprocessing + distributed, I will figure out the design of mixed reading services first then add such feature.

**Comment by ejguan** [comment]
Actually, there is one thing left, which is doc and tutorial for DistributedReadingService.

**Comment by ejguan** [comment]
I will wait until https://github.com/pytorch/pytorch/pull/83741 is landed and released into nightly because it will also use the updated API.

**Comment by ejguan** [comment]
Failing DataPipe tests because the nightly binaries for mac haven't been updated yet.

**Review by NivekT — APPROVED** [review]
LGTM!

nit comment:
IIUC, the difference between this and `_test_distributed_rs` is DL1 vs DL2?

- If so, we can potential remove the duplicate code (not urgent).
- Alternatively, we can just label `_test_distributed_dl` and `_test_distributed_dataloader` (and `elastic_dl`/`elastic_training`) so it will be obvious from the first glance that those two are mostly the same but testing different version of DL.

File: test/test_distributed.py [hunk 3]
```diff
@@ -24,10 +27,25 @@
...
 if not dist.is_available():
+    import sys
+
     print("Distributed not available, skipping tests", file=sys.stderr)
     sys.exit(0)
 
 
+_backends = ["gloo"]
+#  if dist.is_mpi_available():
+#      _backends.append("mpi")
```
[the remark this answers is not in the document]
Comment (ejguan) [reply test/test_distributed.py]: Fixed

**Comment by ejguan** [comment]
> IIUC, the difference between this and `_test_distributed_rs` is DL1 vs DL2?
> 
> * If so, we can potential remove the duplicate code (not urgent).
> * Alternatively, we can just label `_test_distributed_dl` and `_test_distributed_dataloader` (and `elastic_dl`/`elastic_training`) so it will be obvious from the first glance that those two are mostly the same but testing different version of DL.

Fixed

File: torchdata/dataloader2/reading_service.py [hunk 4]
```diff
@@ -235,3 +240,70 @@ def finalize(self) -> None:
...
+            raise RuntimeError("Torch Distributed is required to be initialized")
+        self._world_size = dist.get_world_size()
+        self._rank = dist.get_rank()
+        self._pg = dist.new_group(backend="gloo", timeout=timedelta(seconds=self._timeout))
+        torch.utils.data.graph_settings.apply_sharding(
+            datapipe,
+            self._world_size,
+            self._rank,
+        )
+        datapipe = datapipe.fullsync(self._timeout)
```
Comment (VitalyFedyunin) [inline torchdata/dataloader2/reading_service.py:280]: super nit: Maybe in future check if `fullsync` already in pipeline (some other cascaded RS added it, or user did it manually).

Comment (ejguan) [reply torchdata/dataloader2/reading_service.py:280]: It's a good point. It should be a minimum change with the existing graph functions. Will send a patch shortly.

**Comment by ejguan** [comment]
I will land this PR until PyTorch nightly is updated.
